### ISSUE TITLE
Windows Sonic virtual 7.1 audio support

### ISSUE THREAD

**Issue by S2F0amEgS2xvdmVy** [mpv-player/mpv#7521]
- mpv version: 0.32.0-233-g28ea1ed296 (shinchiro)
- ffmpeg version: git-2020-03-07-cb3c77cfe
- Windows version: 1809 17763.1098
- Audio driver: Microsoft audio driver 10.0.17763.737
- Stereo headphones connected to an audio jack

While attempting to play a 5.1 audio file when the "Windows Sonic for Headphones" feature is enabled in Windows ([image](https://i.ibb.co/PYS9xVm/Sonic.png)), mpv does not see the virtual 7.1 output mode and downmixes the audio to stereo. When using `--audio-exclusive=yes` it is also downsampling to stereo.

Windows Media Player plays the same file with this Sonic mode. Games like Minecraft (Java) and Hearthstone and others also have the Sonic mode enabled ([image](https://i.ibb.co/wBdKCSf/active.png)).

I'm not sure whether I should say this is a bug, as some people might prefer the way mpv downmixes to stereo devices over the virtual surround mode of Windows even if they have it enabled. It is at least quite different to other multi-channel aware applications.

Log file `--no-config`: https://0x0.st/iAtC.txt
Log file `--no-config --audio-exclusive=yes`: https://0x0.st/iAvr.txt
[Backup](https://github.com/mpv-player/mpv/files/4324034/Sonic.zip)
Notable I think:
```
[   0.228][v][ao/wasapi] Trying 5.1 float (32/32 bits) @ 48000hz (shared) -> ok
[   0.228][v][ao/wasapi] Closest match is stereo float (32/32 bits) @ 44100hz
[   0.228][v][ao/wasapi] Accepted as stereo float @ 44100hz -> stereo float (32/32 bits) @ 44100hz (shared)
```

Edit: strangely enough I do get 5.1 output in mpv with the [Fraunhofer test](https://www2.iis.fraunhofer.de/AAC/multichannel.html) (tried the HE-AAC file). Log here: https://0x0.st/iAwS.txt

**Comment by kevmitch** [comment]
So it looks like your system mix format is (44.1khz, stereo, float32 - you may see either 16 or 24 bit in control panel). 

The reason the file you linked works is because it also is 44.1khz.

The reason the file you didn't link did not work is because, since it is 48khz, it does *not* match the system mix format.

So temporary workaround is to set the control panel mix format to match the sample rate of your source material.

The underlying problem is that when the sample rate does match, [`IAudioClient::IsFormatSupported`](https://docs.microsoft.com/en-us/windows/win32/api/audioclient/nf-audioclient-iaudioclient-isformatsupported) returns `S_OK`, and just accepts the format including the channel layout.

When the sample rate does not match, it returns `S_FALSE` indicating that the specified format is not supported, but it returns the "closest match" that should work instead, which is what mpv uses. 

Unfortunately, the closest match is *not* actually the closest match, which would leave the channel layout untouched and only tweak the sample rate. Instead, it just returns the system mix format, which is stereo.

**Comment by garoto** [comment]
I use an APO that does all sorts of audio filtering and manipulation, and I remember having to set `audio-samplerate=48000` in `mpv.conf` so it matched my configured sample rate for my audio device and for the APO configuration to be properly applied.

If this "Windows Sonic" is an APO perhaps setting `--audio-samplerate` to match the sample rate of the device can help.

**Comment by S2F0amEgS2xvdmVy** [comment]
@kevmitch Great thinking! I had factored that out since `--audio-exclusive=yes` picked 48/24 automatically and then it didn't work. But setting the format to 48/24 in the control panel did work.
Then what @garoto wrote works as well. Setting audio-samplerate to the format of the device makes all media multi-channel.

Thanks both. Wonderful thinking!

**Comment by kevmitch** [comment]
@S2F0amEgS2xvdmVy I also meant to mention that exclusive mode is definitely not what you want in this instance. Exclusive mode bypasses all software processing and Windows effects (like Sonic) so that mpv talks directly the audio hardware. Exclusive mode can be useful when matching mpv's output to the hardware's multi-channel capabilities, but for Sonic or anything similar to work, you need Windows to act as an intermediary (i.e., shared mode).

I'm going to reopen this because I think mpv should just work in this instance without the user having to tweak control panel or mpv.conf. I will hopefully get to this this week.

**Comment by kevmitch** [comment]
Fixed by 3aad89829f832091d483589fd578f6ab5b8ba685

**Comment by S2F0amEgS2xvdmVy** [comment]
I finally got around to try it. Works brilliantly. Thanks a lot!
